Infer encode/decode (major changes)

jyn514 opened the pull request:
Assign defaults *after* inferring argument

-  If arguments are specified, easier to infer
-  If not, it's clear that no change was made (as opposed to the user specifying the default)
-  Use dictionary (called 'defaults') to specify
-  Note: filename has to be evaluated second because it depends on input

Quit if no arguments given
Many layers of checks to infer arguments
Decisions made as follows: (assuming encode/decode not passed)

-  editor given        -> decode
-  compresstype given  -> encode
-  compresslevel given -> encode
-  chunklevel given    -> encode
-  filename given      -> encode
-  no output:          -> encode       (why?)
-  no input and interactive  -> decode (why?)
-  input and not interactive -> quit with error
-  input and output          -> quit with error

Other refactoring:
find_editor() is its own function (simplifies logic by a great deal)

jyn514 committed:
Add Github release uploading

Closes https://github.com/charlesdaniels/bitshuffle/issues/11.
Needs testing.

jyn514 committed:
Add pypi metadata

Add version to program
Add Charles as author
Add github page as homepage
Add MANIFEST, setup.cfg

jyn514 committed:
Add build script

Fix typo in setup.py

jyn514 committed:
Improvied editor detection

* Fixed crash if editor cannot be found

* Fix "cannot find editor" if mimeopen is not installed

* added notepad for windows users

* Additionally:
       * check for no editor only if EDITOR is not defined
       * remove useless use of '' as default

Author:    Joshua Nelson <jyn514@gmail.com>

jyn514 committed:
Better exit codes under error conditions

Closes https://github.com/charlesdaniels/bitshuffle/issues/14

jyn514 committed:
Improved do_test & added new test cases

* Refactored do_test to be much easier to add to

* Added function do_test which abstracts most of the
necessary boilerplate

* Added double encode/decode test

* Added tests for python 2 and 3

jyn514 committed:
Infer encode/decode (major changes)

Assign defaults *after* inferring argument
        If arguments are specified, easier to infer
        If not, it's clear that no change was made
                (as opposed to the user specifying the default)
        Use dictionary (called 'defaults') to specify
        Note: filename has to be evaluated second because it depends
              on input

Quit if no arguments given
Many layers of checks to infer arguments
Decisions made as follows: (assuming encode/decode not passed)
        editor given        -> decode
        compresstype given  -> encode
        compresslevel given -> encode
        chunklevel given    -> encode
        filename given      -> encode
        no output:          -> encode       (why?)
        no input and interactive  -> decode (why?)
        input and not interactive -> quit with error
        input and output          -> quit with error

Other refactoring:
find_editor() is its own function (simplifies logic by a great deal)

bx-r0 commented:
Does "Quit if no arguments given" prevent the use of the tool being used with piping?

jyn514 commented:
Only if no arguments are given on the command line. This behavior is specified by https://github.com/charlesdaniels/bitshuffle/issues/1. The tools runs correctly if arguments are specified and input is piped.

jyn514 commented:
You make a good point though- if it's piped, the user won't be looking for help, so we should consider inferring an argument.

Also, I want to refactor the control structure, it has all the tabs because of design decision I never commited

jyn514 committed:
Check if file exists before opening

That is, the file specified by --filename

jyn514 committed:
Refactor smart infer (more)

Ease of reading:
Move smart infer to own function
Move `set defaults` to own function
 - Move said defaults to said function

Control flow:
Use one big if statement for `encode` instead of 4
Exit in one controlled place if cannot infer
Simplify some more logic

TODO:
Add tests. This really needs tests.

Misc:
Remove stderr message when choosing editor

jyn514 committed:
Fix egregious errors

infer_mode returns args
editor changed to args.editor

jyn514 committed:
Print usage by default instead of help

Difference of about 10 lines of terminal.
Prints that --help is a flag, so I doubt it'll
be terribly confusing.

charlesdaniels commented:
> no output: -> encode (why?)

I don't remember specifying this... I can't think of a justification for it; unless the user also specified ``--input``, in which case it makes sense. 

> no input and interactive -> decode (why?)

This only makes sense if standard in is a tty; it means the use is running interactively, and thus that they want their `$VISUAL` to open so they can paste data in.

> input and not interactive -> quit with error

Huh? This would imply the user is using it with a pipeline. For example `bitshuffle --input /some/file | some other script`. One specific use case for this functionality would be where the input needs to come from a named pipe (although the user could easily direct it to stdin.

> input and output -> quit with error

This is not an error condition, it implies that encoded packets are to be read from the input file specified and written to the specified output file. 

> find_editor() is its own function (simplifies logic by a great deal)

This was a good idea. I tried to find a library to do this, but could not. 

> You make a good point though- if it's piped, the user won't be looking for help, so we should consider inferring an argument.

Just because stdin is not a tty does not mean bitshuffle is running in a pipeline (although it is a strong indicator for the purpose of heuristics). Also, the help message is printed on stderr (as far as I know) which would not be captured by the pipe without the user specifically telling the pipe to capture it.

charlesdaniels commented on bitshuffle.py:
```diff
@@ -133,103 +133,96 @@ def encode_file(fhandle, chunksize, compresslevel, compresstype, filename):
         # set to True for infile to be deleted after decoding
         is_tmp = False
-        if stdin.isatty() and args.input is '/dev/stdin':
+        if stdin.isatty() and args.input == '/dev/stdin':
```
As an aside, we should find a portable alternative to `/dev/stdin`, as Windows does not support it, and it's not a POSIX thing to my knowledge. On old school UNIX `/dev/stdin` comes from fdescfs, and from procfs on Linux, but it might not be present on all UNIX variants. Python probably has libraries for dealing with this. Not something that needs to be fixed right now, but should probably be added as an issue.

jyn514 commented:
To clarify, are these the changes you propose to the pull request?

input and no output -> encode
input and output -> decode

As an aside: We should probably find a way to describe input/output as required in the help message, it's very uninformative otherwise.

charlesdaniels commented:
> To clarify, are these the changes you propose to the pull request?
> input and no output -> encode
> input and output -> decode

Yes, that is correct.

> As an aside: We should probably find a way to describe input/output as required in the help message, it's very uninformative otherwise.

Yes, I'm planning to re-work some of the verbiage around this project at some point, there are a number of terms which are used in inconsistent or unclear ways.

jyn514 committed:
Fix conflicting editor options

jyn514 committed:
Check if editor exists

Use existing find_editor function
If nonexistent, print("Editor {editor} not found") and exit with error

charlesdaniels committed:
Merge branch 'staging' of github.com:charlesdaniels/bitshuffle into staging

jyn514 commented:
[Merged this into dev](https://github.com/charlesdaniels/bitshuffle/commit/5bfa519759f1c83298c1d4013a0a956349d2a730); too many changes for it to go straight to master.

charlesdaniels commented:
All of the inference considerations that we discussed have been implemented as of c4a0290

Note that I decided that having `--input` AND `--output` simultaneously should not imply anything special.
